operator/pkg/ciliumendpointslice/endpointslice.go

```diff
@@ -182,24 +183,67 @@ func (c *CiliumEndpointSliceController) Run(ces cache.Indexer, stopCh <-chan str
 // Upon warm boot[restart], Iterate over all CEPs which we got from the api-server
 // and compare it with CEPs packed inside CES.
 // If there are any stale CEPs present in CESs, remove them from their CES.
-func (c *CiliumEndpointSliceController) removeStaleCEPEntries() {
-	log.Info("Remove stale CEP entries in CES")
+// If there are any duplicated CEPs present in CESs, remove all but one trying
+// to keep the CEP with matching identity if it's present.
+func (c *CiliumEndpointSliceController) removeStaleAndDuplicatedCEPEntries() {
+	log.Info("Remove stale and duplicated CEP entries in CES")
+
+	type cepMapping struct {
+		identity int64
+		cesName  string
+	}
+
+	cepsMapping := make(map[string][]cepMapping)
 
 	// Get all CEPs from local datastore
-	staleCEPs := c.Manager.getAllCEPNames()
+	// Map CEP Names to list of whole structure + CES Name
+	for _, ces := range c.Manager.getAllCESs() {
+		for _, cep := range ces.getAllCEPs() {
+			cepName := ces.getCEPNameFromCCEP(&cep)
+			cepsMapping[cepName] = append(cepsMapping[cepName], cepMapping{identity: cep.IdentityID, cesName: ces.getCESName()})
+		}
+	}
 
-	// Remove stale CEP entries present in CES
-	for _, cepName := range staleCEPs {
+	for cepName, mappings := range cepsMapping {
+		storeCep, exists, err := c.ciliumEndpointStore.GetByKey(cepName)
 		// Ignore error from below api, this is added to avoid accidental cep rmeoval from cache
-		if _, exists, err := c.ciliumEndpointStore.GetByKey(cepName); err == nil && exists || err != nil {
+		if err != nil {
 			continue
 		}
-		log.WithFields(logrus.Fields{
-			logfields.CEPName: cepName,
-		}).Debug("Removing stale CEP entry.")
-		c.Manager.RemoveCEPFromCache(cepName, DefaultCESSyncTime)
+		if !exists {
+			// Remove stale CEP entries present in CES
+			for _, mapping := range mappings {
+				log.WithFields(logrus.Fields{
+					logfields.CEPName: cepName,
+				}).Debug("Removing stale CEP entry.")
+				c.Manager.removeCEPFromCES(cepName, mapping.cesName, DefaultCESSyncTime, 0, false)
+			}
+		} else if len(mappings) > 1 {
+			// Remove duplicated CEP entries present in CES
+			found := false
+			cep := storeCep.(*cilium_api_v2.CiliumEndpoint)
+			// Skip first element for now
+			for _, mapping := range mappings[1:] {
```

Comment:
Is it maybe simpler if it goes through the mappings twice? First to find the CEP with the matching ID, and then to remove all but selected (the first CEP or with matching ID).

Reply:
There may be multiple CEPs with matching ID or no CEPs with matching ID.
The code would be something like:
matchingIdIndex := 0
for loop {
   if matches id {
     matchingIdIndex = currentIndex
}
for loop {
  if currentIndex == matchingIdIndex {
    keep
  } else {
    remove
  }
}

I don't think it would be much simpler